File: backend/src/models/external-deal.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum DealSource {
  Skyscanner = 'Skyscanner',
  BookingCom = 'BookingCom',
  Shopify = 'Shopify',
  Amazon = 'Amazon',
  Custom = 'Custom',
}

export interface IExternalDeal extends Document {
  onChainAddress: string;
  oracleAuthority: string;
  source: DealSource;
  externalId: string;
  title: string;
  description: string;
  originalPrice: number;
  discountedPrice: number;
  discountPercentage: number;
  category: string;
  imageUrl: string;
  affiliateUrl: string;
  expiryTimestamp: Date;
  lastUpdated: Date;
  isVerified: boolean;
  verificationCount: number;
  createdAt: Date;
}

const ExternalDealSchema = new Schema<IExternalDeal>(
  {
    onChainAddress: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    oracleAuthority: {
      type: String,
      required: true,
      index: true,
    },
    source: {
      type: String,
      enum: Object.values(DealSource),
      required: true,
      index: true,
    },
    externalId: {
      type: String,
      required: true,
      maxlength: 100,
      index: true,
    },
    title: {
      type: String,
      required: true,
      maxlength: 200,
    },
    description: {
      type: String,
      required: true,
      maxlength: 500,
    },
    originalPrice: {
      type: Number,
      required: true,
    },
    discountedPrice: {
      type: Number,
      required: true,
    },
    discountPercentage: {
      type: Number,
      required: true,
      min: 0,
      max: 100,
    },
    category: {
      type: String,
      required: true,
      maxlength: 50,
      index: true,
    },
    imageUrl: {
      type: String,
      required: true,
      maxlength: 200,
    },
    affiliateUrl: {
      type: String,
      required: true,
      maxlength: 200,
    },
    expiryTimestamp: {
      type: Date,
      required: true,
    },
    lastUpdated: {
      type: Date,
      default: Date.now,
    },
    isVerified: {
      type: Boolean,
      default: false,
    },
    verificationCount: {
      type: Number,
      default: 0,
    },
  },
  {
    timestamps: true,
  }
);

ExternalDealSchema.index({ source: 1, externalId: 1 }, { unique: true });
ExternalDealSchema.index({ category: 1, isVerified: 1 });
ExternalDealSchema.index({ expiryTimestamp: 1 });

export const ExternalDeal = mongoose.model<IExternalDeal>('ExternalDeal', ExternalDealSchema);

File: program/programs/discount-platform/src/state/external_deal.rs
// src/state/external_deal.rs
use anchor_lang::prelude::*;

#[account]
#[derive(InitSpace)]
pub struct ExternalDeal {
    pub oracle_authority: Pubkey,
    pub source: DealSource,
    #[max_len(100)]
    pub external_id: String,
    #[max_len(200)]
    pub title: String,
    #[max_len(500)]
    pub description: String,
    pub original_price: u64,
    pub discounted_price: u64,
    pub discount_percentage: u8,
    #[max_len(50)]
    pub category: String,
    #[max_len(200)]
    pub image_url: String,
    #[max_len(200)]
    pub affiliate_url: String,
    pub expiry_timestamp: i64,
    pub last_updated: i64,
    pub is_verified: bool,
    pub verification_count: u32,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq)]
pub enum DealSource {
    Skyscanner,
    BookingCom,
    Shopify,
    Amazon,
    Custom,
}

// Implement Space trait manually for enum
impl anchor_lang::Space for DealSource {
    const INIT_SPACE: usize = 1;
}

#[account]
#[derive(InitSpace)]
pub struct OracleConfig {
    pub authority: Pubkey,
    pub total_deals_imported: u64,
    #[max_len(10)]
    pub allowed_sources: Vec<DealSource>,
    pub min_verification_count: u32,
    pub update_interval: i64,
}